### DISCUSSION TITLE
`hardlink`: which file is linked to which?

### DISCUSSION THREAD

**Discussion by brianjmurrell** [util-linux/util-linux#3241] [Q&A] [answered]
With `hardlink` can I influence which file is linked to which based on the dir it's found in?  I.e. given two files:

```
181813170 -rw-rw-r--. 30 brian brian 37 Apr 27  2016 /.snapshots/daily.0/pc/home/brian/tmp/FooBar/lib/bencode/BTL.py                                                             
 80746683 -rw-rw-r--. 33 brian brian 37 Apr 27  2016 /.snapshots/weekly.0/pc/home/brian/tmp/FooBar/lib/bencode/BTL.py                                                            
```

I always want the files in `daily.0` to be considered canonical and files in `weekly.0` should be linked to files in `daily.0`.  But I cannot use either of `--{max,min}imize` as there is no consistency to which will have the most/least number of links.

**Comment by karelzak** [comment, marked as answer]
The next release (v2.41)  will support a new option (man page):
```
*-F*, *--prioritize-trees*::
Keep files found in the earliest specified top-level directory if there are multiple identical files in different trees.
For example, *hardlink foo/ bar/* will link files in bar/ to equal files in the foo/ directory.
```

**Reply by brianjmurrell** [reply]
But prior to `--prioritize-trees` it's indeterminate which file is the source and which is the destination of the linking?

**Reply by karelzak** [reply]
The hardlink command should be deterministic for the tree. If you call it twice, you should always get the same result. Files with the same size are ordered by function https://github.com/util-linux/util-linux/blob/master/misc-utils/hardlink.c#L663-L691, and the "greater" file is used as the master (source for link()).

I think we need to add to the hardlink man page description of the algorithm :-)
